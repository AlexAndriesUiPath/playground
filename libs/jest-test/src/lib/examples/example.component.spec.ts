import { ChangeDetectionStrategy } from '@angular/core';
import {
  ComponentFixture,
  TestBed
} from '@angular/core/testing';
import {
  ChildComponent,
  ExampleComponent,
  ExampleService,
  RootService
} from '@playground/examples';
import {
  createMockObject,
  JestTestModule,
  provideMockedObject
} from '@playground/jest-test';
import {
  createComponent,
  PageHelper
} from '@playground/unit-tests';
import { MockComponent } from 'ng-mocks';
import { BehaviorSubject } from 'rxjs';

class ExampleComponentPage extends PageHelper<ExampleComponent> {
  get button(): HTMLElement {
    return this.elementQuery('button');
  }

  get streamLabel(): HTMLElement {
    return this.elementQuery('span');
  }
}

describe('Component(Jest): ExampleComponent', () => {
  let fixture: ComponentFixture<ExampleComponent>;
  let component: ExampleComponent;
  let page: ExampleComponentPage;
  let service: jest.Mocked<ExampleService>;
  let root: jest.Mocked<RootService>;
  let dataStream: BehaviorSubject<{ stream: string }>;

  beforeEach(() => {
    dataStream = new BehaviorSubject<{ stream: string }>({ stream: '' });
    service = createMockObject(ExampleService);
    service.data$ = dataStream.asObservable();
    service.updateStream = jest.fn((stream) => dataStream.next({ stream }));

    TestBed.configureTestingModule({
      imports: [ExampleComponent],
      providers: [
        provideMockedObject(RootService),
      ],
    }).overrideComponent(ExampleComponent, {
      set: {
        changeDetection: ChangeDetectionStrategy.Default,
        imports: [
          JestTestModule,
          MockComponent(ChildComponent)
        ],
        providers: [{
          provide: ExampleService,
          useFactory: () => service
        }]
      }
    });

    root = TestBed.inject(RootService) as jest.Mocked<RootService>;
  });

  beforeEach(() => {
    ({ component, fixture, page } = createComponent(ExampleComponent, ExampleComponentPage));
    fixture.detectChanges();
  });

  it('should match snapshot with initial state', () => {
    expect(service.updateStream).toHaveBeenCalledTimes(1);
    expect(root.init).toHaveBeenCalledTimes(1);
    expect(fixture).toMatchSnapshot();
  });

  it('should call updateStream method', () => {
    const newStreamData = 'new Data';
    component.updateStreamHandler(newStreamData);
    expect(service.updateStream).toHaveBeenCalledWith(newStreamData);

    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should set the `fromClick` value in stream', () => {
    page.button.dispatchEvent(new CustomEvent('click'));

    fixture.detectChanges();

    expect(page.streamLabel.innerHTML).toBe('fromClick');
    expect(fixture).toMatchSnapshot();
  });
});
