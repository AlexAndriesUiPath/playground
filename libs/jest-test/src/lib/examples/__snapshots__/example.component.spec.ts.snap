// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(Jest): ExampleComponent should call updateStream method 1`] = `
<ui-example
  data$={[Function Observable]}
  root={[Function Object]}
  service={[Function Object]}
>
  <span>
    new Data
  </span><ui-child /><img
    alt="alternate text"
    src="#"
  /><button>
    Set 'fromClick'
  </button>
</ui-example>
`;

exports[`Component(Jest): ExampleComponent should match snapshot with initial state 1`] = `
<ui-example
  data$={[Function Observable]}
  root={[Function Object]}
  service={[Function Object]}
>
  <span>
    new stream
  </span><ui-child /><img
    alt="alternate text"
    src="#"
  /><button>
    Set 'fromClick'
  </button>
</ui-example>
`;

exports[`Component(Jest): ExampleComponent should set the \`fromClick\` value in stream 1`] = `
<ui-example
  data$={[Function Observable]}
  root={[Function Object]}
  service={[Function Object]}
>
  <span>
    fromClick
  </span><ui-child /><img
    alt="alternate text"
    src="#"
  /><button>
    Set 'fromClick'
  </button>
</ui-example>
`;
